### ISSUE TITLE
Cognitive complexity metric does not understand newer C++ syntax

### ISSUE THREAD

**Issue by Patschkowski** [SonarOpenCommunity/sonar-cxx#2672]
**Describe the bug**
Code complexity metric is counting r-value references as logical operators and "=default" as switch labels.

**Expected behavior**
Code complexity metric does not count r-value references as logical operators and "=default" as switch labels.

**Screenshots**
![image](https://github.com/SonarOpenCommunity/sonar-cxx/assets/49550034/1ca98955-0f17-4201-8249-d4d2637db6ed)
![image](https://github.com/SonarOpenCommunity/sonar-cxx/assets/49550034/31d25cbf-d8f3-49f4-9099-6561426328f3)

**Desktop (please complete the following information):**
 - OS: Windows
 - SonarQube version: 10.4.1 (build 88267)
 - cxx plugin version: 2.2 (build 626)
 - sonar-scanner version: 5.0.1.3006

**Comment by guwirth** [comment]
Hello @Patschkowski, thanks for the feedback …

**Comment by guwirth** [comment]
Hi @Patschkowski,

most likely the problem is somewhere here: https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-squid/src/main/java/org/sonar/cxx/visitors/CxxCognitiveComplexityVisitor.java

Can you confirm, that it’s a problem with cognitive complexity?

Regards,

**Comment by guwirth** [comment]
Hello @Patschkowski,

there are metrics (Measues) and rule violations (Issues) in SonarQube. The first screenshot you sent seems to show the cognitive complexity metric. The second one is a complexity rule violation. These are two different things created in a different way. Also the visitor verifying the AST is for cognitive and cyclomatic complexity a different one. The reason why I’m asking is to narrow down the problem. 

Regards,

**Comment by Patschkowski** [comment]
Hi @guwirth 

appreciate your explanation. I think it is the Cyclomatic Complexity though:

![image](https://github.com/SonarOpenCommunity/sonar-cxx/assets/49550034/b5e0897a-13b4-4fd8-b453-ac045593f453)

**Comment by guwirth** [comment]
Hi @Patschkowski,

thx for clarification!  Can you copy me the source code of this one template `host` of the screenshot above into this ticket please. Makes testing easier…

Regards,

**Comment by Patschkowski** [comment]
```
#include <memory>

template<typename T>
class host :
  public std::enable_shared_from_this<host<T>> {
public:
  template<typename... U>
  explicit host(U&&... args);

  host() = delete;

  host(const host&) = delete;

  host(host&&)      = default;

  host& operator=(const host&) = delete;

  host& operator=(host&&)      = default;
};
```

**Comment by guwirth** [comment]
Think problem is that [CxxCognitiveComplexityVisitor.java](https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-squid/src/main/java/org/sonar/cxx/visitors/CxxCognitiveComplexityVisitor.java) is using `functionDefinition` as root node. Maybe `functionBody` with child `compoundStatement` would be better?

![grafik](https://github.com/SonarOpenCommunity/sonar-cxx/assets/6077367/5bba15a7-adb9-4d73-af7f-c9ae0c729a00)

**Comment by guwirth** [comment]
Hi @Patschkowski,

from here you can download and test the fix

https://github.com/SonarOpenCommunity/sonar-cxx/actions/runs/8940828967
https://github.com/SonarOpenCommunity/sonar-cxx/actions/runs/8940828967/artifacts/1470960233

Can you give feedback if this is working as expected?

Regards,

**Comment by guwirth** [comment]
@Patschkowski fixed with #2675, you can test it with https://github.com/SonarOpenCommunity/sonar-cxx/releases/tag/latest-snapshot
